tensorflow/compiler/mlir/xla/transforms/legalize_tf.cc

```diff
@@ -1299,6 +1300,193 @@ using ConvertDepthConv2DOp =
     ConvertConvOp<TF::DepthwiseConv2dNativeOp, /*num_spatial_dims=*/2,
                   /*depthwise_conv=*/true>;
 
+class ConvertPadOpStaticDynamic : public OpRewritePattern<TF::PadV2Op> {
+ public:
+  using OpRewritePattern::OpRewritePattern;
+  // Converts a tf.PadV2Op to mhlo.PadOp when input has static shape and
+  // padding values are const. Otherwise, converts the tf.PadV2Op to
+  // mhlo.DynamicPadOp.
+  LogicalResult matchAndRewrite(TF::PadV2Op op,
+                                PatternRewriter& rewriter) const override {
+    auto loc = op.getLoc();
+    auto input = op.input();
+    auto paddings = op.paddings();
+    auto constant_values = op.constant_values();
+    auto input_type = input.getType().dyn_cast<RankedTensorType>();
+    auto paddings_type = paddings.getType().dyn_cast<RankedTensorType>();
```

Comment:
Is this a general assumption you've made, that input graphs have static ranks but dynamic shapes?

Reply:
Yes, We assume rank is always static.

Reply:
We implemented a "static_rank_compilation_cache" in the runtime op. The logics are similar to the xla_compilation_cache, but on a different level (shape->rank) to guarantee that the ranks can always be known at compile time. The compilation cache in theory can support dynamic rank workloads (with multiple compilations) but actually rare to see such workloads anyway. This is why we have such assumptions.
For the specific codes here, the dynamic rank support of this lowering pattern can be regarded as TODO. Maybe we should add some comments here. @azazhu

Reply:
SG.